h2/src/main/org/h2/command/Parser.java

```diff
@@ -7866,9 +7866,29 @@ private Prepared parseSet() {
             readIfEqualOrTo();
             Set command = new Set(session, SetTypes.SCHEMA_SEARCH_PATH);
             ArrayList<String> list = Utils.newSmallArrayList();
+            boolean hasPgCatalog = false;
             do {
-                list.add(readAliasIdentifier());
+                // some PG clients will send single-quoted alias
+                String s = currentTokenType == LITERAL ?
+                        readString() : readAliasIdentifier();
+                if ("$USER".equalsIgnoreCase(s)) {
+                    continue;
+                }
+                if ("PG_CATALOG".equalsIgnoreCase(s)) {
+                    hasPgCatalog = true;
+                }
+                list.add(s);
             } while (readIf(COMMA));
+            // for PostgreSQL compatibility
+            // If "pg_catalog" is not in the path then it will be searched before
+            // searching any of the path items. See
+            // https://www.postgresql.org/docs/8.2/runtime-config-client.html
+            if (!hasPgCatalog) {
+                String pgCatalog = database.sysIdentifier("PG_CATALOG");
+                if (database.findSchema(pgCatalog) != null) {
+                    list.add(0, pgCatalog);
+                }
+            }
```

Comment:
All PostgreSQL-specific logic here should be restricted to PostgreSQL compatibility mode.

Reply:
fixed